### PR TITLE
Gdrive checkpointed connector

### PR DESCRIPTION
**Pull request by evan-danswer**
## Description

Addresses https://linear.app/danswer/issue/DAN-1505/google-drive-checkpointed-connector

- Refactored CheckpointConnector to have better typing
- Implemented Checkpointing for google drive using OAuth and service account credentials
  - Checkpointing is done in stages based on what the user specified they wanted retrieved
  - within each stage, retrieval is now ordered by modified time to allow returning to the stage at the stored time
  - When retrieval reaches the folder crawling stage, checkpointing is based on a set of previously seen parent folder ids. This means we will re-traverse the folder structure to find new docs to index. potential future work could include making this more efficient, but it likely isn't too many extra calls since we exit early from already-seen folders.
  - Slim retrieval modifies a dummy checkpoint in some cases, but this shouldn't cause any issues.
- fixed a parallelism issue for service account retrieval (was submitting a sequence of generators to as_completed)
- added some parallel utilities to support the above with associated tests

## How Has This Been Tested?

So far, a full run of OAuth drive retrieval has worked, with a small number of overlapped docs (the last doc of one indexing batch become the first one retrieved by the next run of that checkpoint). 

Testing checklist (X = done):
X full run with OAuth
- simulated error(s) with OAuth
- full exit halfway through and continue from checkpoint later with OAuth
- a full run with service account
- simulated error(s) with service account
- full exit halfway through and continue from checkpoint later with service account

TODO automated tests:
- unit test for serializing/deserializing drive checkpoint
- unit tests for restarting from various points in the checkpointing process
  - include boundary cases (ended on last doc in a section)

## Backporting (check the box to trigger backport action)

Note: You have to check that the action passes, otherwise resolve the conflicts manually and tag the patches.

- [ ] This PR should be backported (make sure to check that the backport attempt succeeds)
- [ ] [Optional] Override Linear Check

### CODE REVIEW

File: backend/onyx/connectors/google_drive/connector.py [hunk 1]
```diff
@@ -350,15 +375,24 @@ def _impersonate_user_for_retrieval(
...
                 end=end,
+                key=lambda check: user_email
+                + "@"
+                + drive_id,  # completion map keyed by drive id
             )
 
+        # I believe there may be some duplication here,
+        # i.e. if two users have access to the same folder
+        # and are retrieving in parallel.
         remaining_folders = filtered_folder_ids - self._retrieved_ids
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/connectors/google_drive/connector.py]: Duplicate retrievals isn't that bad relative to what it would take to fix this (synchronized retrieved_ids across threads)

File: backend/onyx/connectors/interfaces.py [hunk 2]
```diff
@@ -184,14 +190,21 @@ def handle_event(self, event: Any) -> GenerateDocumentsOutput:
         raise NotImplementedError
 
 
-class CheckpointConnector(BaseConnector):
+CT = TypeVar("CT", bound=ConnectorCheckpoint)
+# TODO: find a reasonable way to parameterize the return type of the generator
+CheckpointOutput: TypeAlias = Generator[
+    Document | ConnectorFailure, None, ConnectorCheckpoint
+]
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/connectors/interfaces.py:199]: Yeah, I also wish Type Aliases and Generics worked like that. They don't, though.

File: backend/onyx/connectors/interfaces.py [hunk 3]
```diff
@@ -212,3 +225,10 @@ def load_from_checkpoint(
         ```
         """
         raise NotImplementedError
+
+    # Ideally return typeshould be CT, but that's not possible if
+    # we want to override build_dummy_checkpoint and have BaseConnector
+    # return a base ConnectorCheckpoint
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/connectors/interfaces.py]: 🐐

File: backend/onyx/connectors/models.py [hunk 4]
```diff
@@ -198,18 +197,13 @@ class IndexAttemptMetadata(BaseModel):
...
-
     def __str__(self) -> str:
         """String representation of the checkpoint, with truncation for large checkpoint content."""
         MAX_CHECKPOINT_CONTENT_CHARS = 1000
 
-        content_str = json.dumps(self.checkpoint_content)
+        content_str = self.model_dump_json()
         if len(content_str) > MAX_CHECKPOINT_CONTENT_CHARS:
             content_str = content_str[: MAX_CHECKPOINT_CONTENT_CHARS - 3] + "..."
         return f"ConnectorCheckpoint(checkpoint_content={content_str}, has_more={self.has_more})"
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/connectors/models.py]: Good point, I kinda autopiloted here

File: backend/onyx/connectors/google_drive/models.py [hunk 5]
```diff
@@ -20,3 +24,75 @@ class GDriveMimeType(str, Enum):
...
+    completion_map: ThreadSafeDict[str, SecondsSinceUnixEpoch] = ThreadSafeDict()
+
+    # cached version of the drive and folder ids to retrieve
+    drive_ids_to_retrieve: list[str] | None = None
+    folder_ids_to_retrieve: list[str] | None = None
+
+    # cached user emails
+    user_emails: list[str] | None = None
+
+    # @field_serializer("completion_map")
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/models.py]: nit: remove if unused

File: backend/onyx/connectors/interfaces.py [hunk 6]
```diff
@@ -186,14 +192,21 @@ def handle_event(self, event: Any) -> GenerateDocumentsOutput:
         raise NotImplementedError
 
 
-class CheckpointConnector(BaseConnector):
+CT = TypeVar("CT", bound=ConnectorCheckpoint)
+# TODO: find a reasonable way to parameterize the return type of the generator
+CheckpointOutput: TypeAlias = Generator[
```
Comment (Weves) [inline backend/onyx/connectors/interfaces.py:197]: should we remove/comment out if not usable due to generic annoying-ness?

File: backend/onyx/utils/threadpool_concurrency.py [hunk 7]
```diff
@@ -190,3 +308,30 @@ def wait_on_background(task: TimeoutThread[R]) -> R:
...
+        future_to_index: dict[Future[tuple[int, R | None]], int] = {
+            executor.submit(_next_or_none, i, g): i for i, g in enumerate(gens)
+        }
+
+        next_ind = len(gens)
+        while future_to_index:
+            done, _ = wait(future_to_index, return_when=FIRST_COMPLETED)
+            for future in done:
+                ind, result = future.result()
+                print(ind, result)
```
Comment (Weves) [inline backend/onyx/utils/threadpool_concurrency.py]: nit: remove `print`

File: backend/tests/daily/connectors/google_drive/consts_and_utils.py [hunk 8]
```diff
@@ -202,3 +204,13 @@ def assert_retrieved_docs_match_expected(
         retrieved=valid_retrieved_texts,
     )
     assert expected_file_texts == valid_retrieved_texts
+
+
+def load_all_docs(connector: GoogleDriveConnector) -> list[Document]:
```
Comment (Weves) [inline backend/tests/daily/connectors/google_drive/consts_and_utils.py:218]: this only loads all docs from the first checkpoint if I understand the function correctly? If so, should probably be re-named

File: backend/onyx/connectors/google_drive/connector.py [hunk 9]
```diff
@@ -361,25 +395,27 @@ def _impersonate_user_for_retrieval(
...
-        elif self.include_shared_drives:
-            drive_ids_to_retrieve = all_drive_ids
+        if checkpoint.completion_stage == DriveRetrievalStage.START:
+            checkpoint.completion_stage = DriveRetrievalStage.USER_EMAILS
+
+        if checkpoint.completion_stage == DriveRetrievalStage.USER_EMAILS:
+            all_org_emails: list[str] = self._get_all_user_emails()
+            if not is_slim:
+                checkpoint.user_emails = all_org_emails
+            checkpoint.completion_stage = DriveRetrievalStage.DRIVE_IDS
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py:532]: we should `return` checkpoint here to save our progress

File: backend/onyx/connectors/google_drive/file_retrieval.py [hunk 10]
```diff
@@ -243,16 +287,22 @@ def get_all_files_for_oauth(
     yield from execute_paginated_retrieval(
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/file_retrieval.py:290]: to properly checkpoint this, we should return back a new checkpoint every X calls?

Comment (Weves) [reply backend/onyx/connectors/google_drive/file_retrieval.py:290]: similar comment for most of the other flows

Comment (Weves) [reply backend/onyx/connectors/google_drive/file_retrieval.py:290]: nvm on ^, looking into the short-circuited approach we discussed

File: backend/onyx/connectors/google_drive/connector.py [hunk 11]
```diff
@@ -499,46 +579,163 @@ def _manage_oauth_retrieval(
...
+                            executor.submit(convert_func, file) for file in files_batch
+                        ]
+                        documents = []
+                        for future in as_completed(futures):
+                            try:
+                                doc = future.result()
+                                if doc is not None:
+                                    documents.append(doc)
+                            except Exception as e:
+                                logger.error(f"Error converting file: {e}")
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: we should probably yield a failure here? E.g. if a single document fails to be downloaded

File: backend/onyx/connectors/google_drive/connector.py [hunk 12]
```diff
@@ -499,46 +579,163 @@ def _manage_oauth_retrieval(
...
+                            yield documents
+                            batches_complete += 1
+                        files_batch = []
+
+                        if batches_complete > BATCHES_PER_CHECKPOINT:
+                            checkpoint.retrieved_ids = list(self._retrieved_ids)
+                            return  # create a new checkpoint
+
+                # Process any remaining files
+                if files_batch:
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py:976]: same note on yielding a failure below

File: backend/onyx/connectors/google_drive/connector.py [hunk 13]
```diff
@@ -553,40 +750,52 @@ def _extract_docs_from_google_drive(
...
-                        logger.error(f"Error converting file: {e}")
-
-                if documents:
-                    yield documents
-
-    def load_from_state(self) -> GenerateDocumentsOutput:
-        try:
-            yield from self._extract_docs_from_google_drive()
         except Exception as e:
+            logger.exception(f"Error extracting documents from Google Drive: {e}")
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py:1005]: if we reach here, it's likely a terminal / massive failure, right? The current implementation, would could mark the index attempt as a "success with failures" despite only completing a few docs out of millions?

If so, we should probably just raise the exception regardless, and error out loudly

File: backend/tests/daily/connectors/google_drive/consts_and_utils.py (hunk 8 above)
Comment (evan-onyx) [reply backend/tests/daily/connectors/google_drive/consts_and_utils.py:218]: It's a testing function and it doesn't take a checkpoint as a parameter so I feel like it's pretty self-explanatory? Would you want something like load_all_docs_from_empty or _from_first_checkpoint?

File: backend/onyx/connectors/interfaces.py (hunk 6 above)
Comment (evan-onyx) [reply backend/onyx/connectors/interfaces.py:197]: CT is still being used, I figure this is worth leaving in for when python typing libraries get better

File: backend/onyx/connectors/google_drive/connector.py [hunk 14]
```diff
@@ -238,6 +241,25 @@ def load_credentials(self, credentials: dict[str, Any]) -> dict[str, str] | None
 
         return new_creds_dict
 
+    def _checkpoint_yield(
+        self,
+        drive_files: Iterator[GoogleDriveFileType],
+        checkpoint: GoogleDriveCheckpoint,
+        key: Callable[
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: prefer a more descriptive name here (`key` is not clear if it's a function or an actual key + what exactly it does)

File: backend/onyx/connectors/google_drive/connector.py [hunk 15]
```diff
@@ -330,8 +353,11 @@ def _impersonate_user_for_retrieval(
                 service=drive_service,
                 update_traversed_ids_func=self._update_traversed_parent_ids,
                 is_slim=is_slim,
+                checkpoint=checkpoint,
                 start=start,
                 end=end,
+                key=lambda check: user_email
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: is there a realistic reason why someone would call this function with a different key? If so, this flexibility feels unnecessary and probably makes it harder to understand what's going on in `get_all_files_in_my_drive`. E.g. the reader of `get_all_files_in_my_drive` has to look at where it was called to understand the expected behavior

File: backend/onyx/connectors/google_drive/connector.py [hunk 16]
```diff
@@ -341,15 +367,23 @@ def _impersonate_user_for_retrieval(
                 service=drive_service,
                 drive_id=drive_id,
                 is_slim=is_slim,
+                checkpoint=checkpoint,
                 update_traversed_ids_func=self._update_traversed_parent_ids,
                 start=start,
                 end=end,
+                key=lambda check: user_email
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: same comment here as above

File: backend/onyx/connectors/google_drive/connector.py (hunk 9 above)
Comment (Weves) [reply backend/onyx/connectors/google_drive/connector.py:532]: as a general rule, I think it would be good to return after each stage. It's slightly safer (e.g. we don't lose the progress of fetching all user emails if the first call of the next step fails)

**Review by Weves — COMMENTED** [review]
As a general rule, we prefer to not have functions modify objects in place and have the caller's functionality depend on these "hidden" modifications. It's a much clearer contract (and as a result, hopefully less bugs) if a function's behavior is clearly defined by its input and output (I'm sure you're already familiar with all the theory/rationale behind it, but just re-iterating the key point imo). 

All of these functions appear to modify the checkpoint in place in a way that effects the caller. Instead, would it be possible to yield/return back statuses which allow the parent function to modify the deep copied checkpoint -> none of the dependence on in-place modifications of objects passed in.

File: backend/tests/daily/connectors/google_drive/consts_and_utils.py (hunk 8 above)
Comment (Weves) [reply backend/tests/daily/connectors/google_drive/consts_and_utils.py:218]: well it doesn't actually load all docs, right? It loads from "Zero Checkpoint" to "Checkpoint 1". So it could be a tiny subset of all docs to load

File: backend/onyx/agents/agent_search/deep_search/main/nodes/extract_entities_terms.py [hunk 17]
```diff
@@ -106,6 +106,7 @@ def extract_entities_terms(
         first_bracket = cleaned_response.find("{")
         last_bracket = cleaned_response.rfind("}")
         cleaned_response = cleaned_response[first_bracket : last_bracket + 1]
+        cleaned_response += "g8o723dg87db"
```
Comment (Weves) [inline backend/onyx/agents/agent_search/deep_search/main/nodes/extract_entities_terms.py]: what's going on here 🧐

File: backend/onyx/connectors/google_drive/connector.py [hunk 18]
```diff
@@ -287,22 +342,96 @@ def get_all_drive_ids(self) -> set[str]:
...
+            if drive_id in self._retrieved_ids
+            else DriveIdStatus.AVAILABLE
+            for drive_id in drive_ids
+        }
+
+        def _get_available_drive_id(
+            processed_ids: set[str], thread_id: str
+        ) -> tuple[str | None, bool]:
+            found_future_work = False
+            print(f"{thread_id} drive_id_status: {drive_id_status}")
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: nit: remove print

File: backend/onyx/connectors/google_drive/connector.py [hunk 19]
```diff
@@ -113,7 +122,54 @@ def _clean_requested_drive_ids(
...
+
+class DriveIdStatus(Enum):
+    AVAILABLE = "available"
+    IN_PROGRESS = "in_progress"
+    FINISHED = "finished"
+
+
+class DriveIdServer:
+    """
+    Returns an iterator over drive_ids.
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: How does this actually work? I don't see an `__iter__` implementation

Comment (Weves) [reply backend/onyx/connectors/google_drive/connector.py]: seems unused? If so, should remove

File: backend/onyx/connectors/google_drive/connector.py [hunk 20]
```diff
@@ -287,22 +342,96 @@ def get_all_drive_ids(self) -> set[str]:
...
         logger.info(f"Impersonating user {user_email}")
-
+        curr_stage = checkpoint.completion_map.get(user_email, None)
+        resuming = True
+        if curr_stage is None:
+            curr_stage = StageCompletion(
+                stage=DriveRetrievalStage.MY_DRIVE_FILES,
+                completed_until=start or 0,
+            )
+            checkpoint.completion_map[user_email] = curr_stage
```
Comment (Weves) [inline backend/onyx/connectors/google_drive/connector.py]: this looks like we still have some of the behavior from before? E.g. the correct functioning of this connector depends on the lower level functions modifying the object in place?

**Review by Weves — COMMENTED** [review]
We may want to spend some time cleaning this up. This single class is a bit massive with a very deep chain with somewhat redundant sounding functions (e.g. `load_from_checkpoint` -> `_extract_docs_from_google_drive` -> `_fetch_drive_items` -> `_checkpointed_retrieval` -> `_manage_service_account_retrieval` -> `_impersonate_user_for_retrieval` -> `get_all_files_in_my_drive`.

Maybe some of the functions can be flatted and/or the service account retrieval/oauth retrieval can be split into separate classes (or functions) or maybe different files.

Not for now since we do want to get this out ASAP, but something to think about in the future.

File: backend/onyx/connectors/google_drive/connector.py (hunk 19 above)
Comment (evan-onyx) [reply backend/onyx/connectors/google_drive/connector.py]: whoops yeah I switched to a different approach, this was just left over

File: backend/onyx/connectors/google_drive/connector.py (hunk 20 above)
Comment (evan-onyx) [reply backend/onyx/connectors/google_drive/connector.py]: Correct functionality doesn't depend on this at all, I'll just delete this line

**Review by Weves — APPROVED** [review]
